enhancements/machine-api/machine-deletion-hooks.md

```diff
@@ -0,0 +1,424 @@
+as set out below.
+
+#### Machine API waits on all Finalizers
+
+In this scenario, the Machine API would not start draining/removing a Machine until all finalizers
+other than its own one have been removed. This would allow future controllers to manage the removal
+of the Machine without having Machine API have to know anything about their specific implementation.
+However, there could be a scenario where another controller also needs to wait for other finalizers
+to be removed, in such a case, unless the Machine API or this new controller are aware of each other
+there would be a deadlock created.
+
+Relying on all finalizers being removed before draining/removing could become fragile in the future
+due to the above, and as such is not a preferred solution.
+
+#### Machine API waits on some Finalizers
+
+In this scenario, Machine API would have a list of known finalizers that would prevent it from taking
+certain actions. For example a known etcd-quorum finalizer could prevent the Machine controller from
+draining the Machine.
+
+This creates a tight coupling between the Machine controller and other components within OpenShift.
+The upstream solution today, matches the proposal set out in this document. As we are experimenting
+with Cluster API in HyperShift, Central Machine Management and for future IPI clusters, it is
+preferred to not create a tight coupling with these components and to copy the upstream solution to
+allow these mechanisms to be transparently honoured by both MAPI and CAPI going forward.
+
+#### Machine API waits on Finalizers based on an annotation
+
+In this scenario, the blocking operator would use an annotation to instruct the Machine controller
+of a particular finalizer it should wait on. This solution allows the third party operator to opt
+in to having their cleanup logic block the removal of the the Machine, without having to have the
+Machine controller aware of the finalizer ahead of time.
+
+As this approach includes the use of annotations to prevent the Machine controller from taking any
+action, it is very similar to the approach described in the proposal above.
+Since any deletion hook annotation prevents the Mahcine controller from removing its own finalizer,
+this secondary annotation + finalizer combination doesn't add much value over the existing proposal.
+As it also means we deviate from the upstream solution, and may have to account for this in the
+future with additional complexity, this is not a preferred solution.
+
+### Status Field
+Harder for users to modify or set hooks during machine creation.
+How would a user remove a hook if a controller that is supposed to remove it is misbehaving?
+We’d probably need an annotation like ‘skip-hook-xyz’ or similar and that seems redundant
+to just using annotations in the first place.
+
+### Spec Field
+We probably don’t want other controllers dynamically adding and removing spec fields on an object.
```

Comment:
I can't tell if this is a concern about granting the permission to another controller or a claim that server-side-apply doesn't allow for "I want this field to exist in spec" (I think it does).

Reply:
setting annotations requires spec update powers, so I don't think you're gaining anything rbac-wise.

Reply:
> I'll expand this once I've double checked the upstream conversation that happened around this, but as I mentioned in another comment, I think this was more about trying to avoid mixing controller owned fields into the spec which is normally owned by a user. Will get back to this

I'm interested in seeing a link from here.  Given that server-side-apply is based on per-field ownership, a subdivided spec is tractable.

Reply:
Looking through the upstream conversation on the enhancement PR (I haven't found the original google doc yet): 

Part of the reason to use annotations is that the keys have to be unique, so it's less likely that there would be conflicts between different controllers who are dealing with the hooks. (https://github.com/kubernetes-sigs/cluster-api/pull/3132#discussion_r446360507)

Another comment specifically on this title is leading to a conversation about possibly making this a spec field in a future iteration, though I don't think anything has been actioned from that (https://github.com/kubernetes-sigs/cluster-api/pull/3132#discussion_r448618893)

Some more concern about the fact that using annotations is unversioned later in the thread, but agreement that this is ok for an alpha API. Interestingly, the CAPI project is looking to move to beta in the next few weeks, there has been no further conversation around this annotation API, I'll poke them to remind them. (https://github.com/kubernetes-sigs/cluster-api/pull/3132#issuecomment-661104241)

If we can convince the upstream community to revisit how this mechanism works and then it moves to a spec field, then that's great, I think that would work well for us, I'm just a little concerned about the timing here. I'd like to provide _something_ for the control plane scaling project to use, preferably with the same API as CAPI so that there's an easy conversion in the future. Perhaps we can implement the annotations temporarily and then move to a spec before we cut a release, assuming we can upstream consensus 🤔

Reply:
This still isn't clear to me.  Ownership in the spec can be separated, it's what server-side-aply is for.  In addition, the permissions required to modify annotations are identical to permissions required to modify spec, so I don't see a benefit permissions-wise.

Duplicates are prevented by setting the appropriate list merge key and enforcing as we do for other fields.

Conflicts can be managed via server-side apply, patches, or (default) conditional updates.  Using a spec field increases expressivity.

Using an annotation for evaluation guarantees that there will be a migration later when a real field is developed.